File: bot/src/providers/jito-client.ts
import {
  PublicKey,
  SystemProgram,
  TransactionInstruction,
  VersionedTransaction,
} from "@solana/web3.js";
import type pino from "pino";
import bs58 from "bs58";

// Jito tip accounts — pick one at random to reduce contention
const JITO_TIP_ACCOUNTS = [
  "96gYZGLnJYVFmbjzopPSU6QiEV5fGqZNyN9nmNhvrZU5",
  "HFqU5x63VTqvQss8hp11i4wVV8bD44PvwucfZ2bU7gRe",
  "Cw8CFyM9FkoMi7K7Crf6HNQqf4uEMzpKw6QNghXLvLkY",
  "ADaUMid9yfUytqMBgopwjb2DTLSokTSzL1zt6iGPaS49",
  "DfXygSm4jCyNCybVYYK6DwvWqjKee8pbDmJGcLWNDXjh",
  "ADuUkR4vqLUMWXxW9gh6D6L8pMSawimctcNZ5pGwDcEt",
  "DttWaMuVvTiduZRnguLF7jNxTgiMBZ1hyAumKUiL2KRL",
  "3AVi9Tg9Uo68tJfuvoKvqKNWKkC5wPdSSdeBnizKZ6jT",
];

/** Jito block engine endpoints by region */
export const JITO_ENDPOINTS = {
  default: "https://mainnet.block-engine.jito.wtf",
  ny: "https://ny.mainnet.block-engine.jito.wtf",
  amsterdam: "https://amsterdam.mainnet.block-engine.jito.wtf",
  frankfurt: "https://frankfurt.mainnet.block-engine.jito.wtf",
  tokyo: "https://tokyo.mainnet.block-engine.jito.wtf",
  slc: "https://slc.mainnet.block-engine.jito.wtf",
} as const;

export type JitoRegion = keyof typeof JITO_ENDPOINTS;

export interface JitoBundleResult {
  bundleId: string;
}

export interface JitoBundleStatus {
  bundleId: string;
  status: "Invalid" | "Pending" | "Failed" | "Landed";
  landedSlot?: number;
  err?: unknown;
}

export class JitoClient {
  private endpoint: string;
  private logger: pino.Logger;

  constructor(region: JitoRegion, logger: pino.Logger) {
    this.endpoint = JITO_ENDPOINTS[region];
    this.logger = logger;
    this.logger.info(
      { endpoint: this.endpoint, region },
      "Jito client initialized"
    );
  }

  /** Get a random tip account to distribute load */
  getRandomTipAccount(): PublicKey {
    const idx = Math.floor(Math.random() * JITO_TIP_ACCOUNTS.length);
    return new PublicKey(JITO_TIP_ACCOUNTS[idx]);
  }

  /** Build a tip instruction (SOL transfer to Jito tip account) */
  buildTipInstruction(
    payer: PublicKey,
    tipLamports: number
  ): TransactionInstruction {
    const tipAccount = this.getRandomTipAccount();
    this.logger.debug(
      { tipAccount: tipAccount.toBase58(), tipLamports },
      "Jito tip target"
    );
    return SystemProgram.transfer({
      fromPubkey: payer,
      toPubkey: tipAccount,
      lamports: tipLamports,
    });
  }

  /**
   * Send a single transaction via Jito's sendTransaction endpoint.
   * Simpler than a full bundle when you only have one transaction.
   */
  async sendTransaction(tx: VersionedTransaction): Promise<string> {
    const serialized = bs58.encode(tx.serialize());

    const res = await fetch(`${this.endpoint}/api/v1/transactions`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        jsonrpc: "2.0",
        id: 1,
        method: "sendTransaction",
        params: [serialized, { encoding: "base58" }],
      }),
    });

    const json = await res.json() as { result?: string; error?: { code: number; message: string } };
    if (json.error) {
      throw new Error(`Jito sendTransaction failed: ${json.error.message}`);
    }

    this.logger.info(
      { signature: json.result },
      "Transaction sent via Jito"
    );
    return json.result!;
  }

  /**
   * Send a bundle of transactions (up to 5) to Jito block engine.
   * Transactions execute atomically in the order provided.
   * The last transaction MUST include a tip to a Jito tip account.
   */
  async sendBundle(
    transactions: VersionedTransaction[]
  ): Promise<JitoBundleResult> {
    if (transactions.length === 0 || transactions.length > 5) {
      throw new Error(
        `Bundle must contain 1-5 transactions, got ${transactions.length}`
      );
    }

    const serialized = transactions.map((tx) => bs58.encode(tx.serialize()));

    const res = await fetch(`${this.endpoint}/api/v1/bundles`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        jsonrpc: "2.0",
        id: 1,
        method: "sendBundle",
        params: [serialized],
      }),
    });

    const json = await res.json() as { result?: string; error?: { code: number; message: string } };
    if (json.error) {
      throw new Error(`Jito sendBundle failed: ${json.error.message}`);
    }

    const bundleId = json.result!;
    this.logger.info(
      { bundleId, txCount: transactions.length },
      "Bundle sent to Jito"
    );

    return { bundleId };
  }

  /** Check the status of a previously submitted bundle */
  async getBundleStatus(bundleId: string): Promise<JitoBundleStatus> {
    const res = await fetch(`${this.endpoint}/api/v1/bundles`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        jsonrpc: "2.0",
        id: 1,
        method: "getBundleStatuses",
        params: [[bundleId]],
      }),
    });

    const json = await res.json() as {
      result?: { value: Array<{ bundle_id: string; status: string; slot?: number; err?: unknown }> };
      error?: { code: number; message: string };
    };

    if (json.error) {
      throw new Error(
        `Jito getBundleStatuses failed: ${json.error.message}`
      );
    }

    const statuses = json.result?.value ?? [];
    if (statuses.length === 0) {
      return { bundleId, status: "Pending" };
    }

    const s = statuses[0];
    return {
      bundleId: s.bundle_id,
      status: s.status as JitoBundleStatus["status"],
      landedSlot: s.slot,
      err: s.err,
    };
  }
}

File: pybot/jito_client.py
"""Jito block engine client — sends transactions/bundles for MEV-competitive landing."""

import random
import httpx
import base58
from loguru import logger
from solders.pubkey import Pubkey
from solders.system_program import transfer, TransferParams
from solders.instruction import Instruction
from solders.transaction import VersionedTransaction

JITO_TIP_ACCOUNTS = [
    "96gYZGLnJYVFmbjzopPSU6QiEV5fGqZNyN9nmNhvrZU5",
    "HFqU5x63VTqvQss8hp11i4wVV8bD44PvwucfZ2bU7gRe",
    "Cw8CFyM9FkoMi7K7Crf6HNQqf4uEMzpKw6QNghXLvLkY",
    "ADaUMid9yfUytqMBgopwjb2DTLSokTSzL1zt6iGPaS49",
    "DfXygSm4jCyNCybVYYK6DwvWqjKee8pbDmJGcLWNDXjh",
    "ADuUkR4vqLUMWXxW9gh6D6L8pMSawimctcNZ5pGwDcEt",
    "DttWaMuVvTiduZRnguLF7jNxTgiMBZ1hyAumKUiL2KRL",
    "3AVi9Tg9Uo68tJfuvoKvqKNWKkC5wPdSSdeBnizKZ6jT",
]

JITO_ENDPOINTS = {
    "default": "https://mainnet.block-engine.jito.wtf",
    "ny": "https://ny.mainnet.block-engine.jito.wtf",
    "amsterdam": "https://amsterdam.mainnet.block-engine.jito.wtf",
    "frankfurt": "https://frankfurt.mainnet.block-engine.jito.wtf",
    "tokyo": "https://tokyo.mainnet.block-engine.jito.wtf",
    "slc": "https://slc.mainnet.block-engine.jito.wtf",
}


class JitoClient:
    def __init__(self, region: str = "default"):
        self.endpoint = JITO_ENDPOINTS.get(region, JITO_ENDPOINTS["default"])
        self._client = httpx.AsyncClient(timeout=10.0)
        logger.info(f"Jito client initialized: {self.endpoint}")

    async def close(self):
        await self._client.aclose()

    def get_random_tip_account(self) -> Pubkey:
        return Pubkey.from_string(random.choice(JITO_TIP_ACCOUNTS))

    def build_tip_instruction(self, payer: Pubkey, tip_lamports: int) -> Instruction:
        tip_account = self.get_random_tip_account()
        # solders transfer returns an Instruction directly
        return transfer(TransferParams(
            from_pubkey=payer,
            to_pubkey=tip_account,
            lamports=tip_lamports,
        ))

    async def send_transaction(self, tx: VersionedTransaction) -> str:
        serialized = base58.b58encode(bytes(tx)).decode()
        resp = await self._client.post(
            f"{self.endpoint}/api/v1/transactions",
            json={
                "jsonrpc": "2.0",
                "id": 1,
                "method": "sendTransaction",
                "params": [serialized, {"encoding": "base58"}],
            },
        )
        data = resp.json()
        if "error" in data and data["error"]:
            raise Exception(f"Jito sendTransaction failed: {data['error'].get('message', data['error'])}")
        sig = data.get("result", "")
        logger.info(f"Tx sent via Jito: {sig}")
        return sig

    async def send_bundle(self, txs: list[VersionedTransaction]) -> str:
        if not 1 <= len(txs) <= 5:
            raise ValueError(f"Bundle must contain 1-5 txs, got {len(txs)}")
        serialized = [base58.b58encode(bytes(tx)).decode() for tx in txs]
        resp = await self._client.post(
            f"{self.endpoint}/api/v1/bundles",
            json={
                "jsonrpc": "2.0",
                "id": 1,
                "method": "sendBundle",
                "params": [serialized],
            },
        )
        data = resp.json()
        if "error" in data and data["error"]:
            raise Exception(f"Jito sendBundle failed: {data['error'].get('message', data['error'])}")
        bundle_id = data.get("result", "")
        logger.info(f"Bundle sent via Jito: {bundle_id} ({len(txs)} txs)")
        return bundle_id
